### ISSUE TITLE
Need a way to set a global PyPI repo URL for all project manually.

### ISSUE THREAD

**Issue by dev-techmoe** [python-poetry/poetry#2940]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

Hello everyone, I'm a python developer live in Mainland China. As well known the network connection is so bad in my country that we cannot connect to the official PyPI repo directly sometimes. So most of us usually some PyPI mirrors in our region (like https://mirrors.aliyun.com/pypi/simple/, it's deployed by AlibabaCloud) to get a fast and stabilizing network connection. PIP gives us a way to set the index globally using `pip config set global.index-url `. But I cannot find related config items in Poetry.
Okay, I know what you want to say, I know I can set repo info in my `pyproject.toml` like this.
```toml
[[tool.poetry.source]]
name = "aliyun"
url = "https://mirrors.aliyun.com/pypi/simple/"
default = true
```
But it will cause some tricky problems. If I have some workmates in other countries, they don't need repo mirrors to speed up service, even it will make them cannot use PyPI repo (thanks to stupid GFW), or we need to use Github Action as our CI service, they all don't need this config item. **So I think the PyPI repo URL should not as a project config in my project repo**, we need a way to set in our dev environment and use it for all projects. Thank you.

**Comment by gaoming714** [comment]
I see what you mean, there should really be a global control, or I hope the project leader can use even pip configuration.
Because I pip to download when it is convenient to use pqi to control repo.

Is there some one can answer it, make poetry use pip repo.

Maybe you need someway to connect google, you know it.

**Comment by abn** [comment]
@dev-techmoe I can definitely appreciate your situation and I do agree that there are various cases where you would want to use a mirror instead of public pypi index.

Poetry has some hard-coded assumptions about PyPI. This means taht simply setting a url for PyPI will not work. The assumptions include (but not limited to), the use of the JSON endpoint to get package metadata, There is no guarentee that this exists for all mirrors. We need to,

1. Allow for poetry to be confired with first-class knowledge of PyPI mirrors.
2. Improve legacy repository support to ensure that we detect "true" legacy repos (ie, without json endpoints).
3. With these changes, we should be able to safely configure mirror urls and identify if they are warehouse compatible and be smart about what endpoints we use.

For this issue we should keep generic environment specific overrides out of scope; but only focus on supporting PyPI mirrors for an environment.

**Comment by teruokun** [comment]
So, I agree that solving for this use-case is blocked on the above (supporting full Pypi mirrors or repositories based on full pypi mirrors that don't implement the json apis), but it's actually a different feature.

I actually also have this use-case, but it's because we have multiple private repositories that we actually want to publish to as they have specific use-cases as we want their ecosystems to be independent of one-another and we want the build logic to all be the same and use the same underlying code repository, but our build logic would do some environment setup before running `poetry` directly.  If it could be done as a part of 'env' or even settable as an option to `install` or `add` on the command line, this would make life much simpler.

**Comment by abn** [comment]
@teruokun I think your usecase can be handled in otherways. For example;

```sh
export POETRY_REPOSITORIES_SOMEWHERE_URL=https://somewhere.com/legacy
export POETRY_HTTP_BASIC_SOMEWHERE_USERNAME=username
export POETRY_HTTP_BASIC_SOMEWHERE_PASSWORD=password
poetry publish --build -r somewhere
```

Alternatively, once #3068 is done, you could do something like this.
```sh
poetry publish --build -r https://somewhere.org/legacy -u username -p password
```

**Comment by NateScarlet** [comment]
Duplicate of #1632

**Comment by finswimmer** [comment]
Thanks for pointing to the duplications @NateScarlet.

**Comment by Felix-neko** [comment]
https://github.com/python-poetry/poetry/pull/4944 -- looks like it's implemented here.
